File: core/src/vct/col/ast/ParallelBarrier.java
package vct.col.ast;

import java.util.EnumSet;

public class ParallelBarrier extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  public enum Fence {
    /** fence global memory */
    Global,
    /** fence work group memory */
    Local
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 

  public final Contract contract;
  
  public final EnumSet<Fence> fences;
  
  public final BlockStatement body;
  
  public ParallelBarrier(Contract contract, EnumSet<Fence> fences, BlockStatement body){
    this.contract=contract;
    this.fences=EnumSet.copyOf(fences);
    this.body=body;
  }

}

File: vercors/src/main/java/vct/col/ast/stmt/composite/ParallelBarrier.scala
package vct.col.ast.stmt.composite

import java.util.ArrayList

import vct.col.ast.generic.ASTNode
import vct.col.ast.stmt.decl.Contract
import vct.col.ast.util.ASTVisitor
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

class ParallelBarrier (val label:String, val contract:Contract, private[this] val fences:ArrayList[String], val body:BlockStatement) extends ASTNode with VisitorHelper {
  val invs = new ArrayList[String](fences)
    
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))

  override def debugTreeChildrenFields(): Iterable[String] = Seq("contract", "body")
  override def debugTreePropertyFields(): Iterable[String] = Seq("label", "fences")
}
